python/strings_udf/strings_udf/__init__.py

```diff
@@ -14,8 +14,47 @@
 
 logger = get_logger()
 
-# tracks the version of CUDA used to build the c++ and PTX components
-STRINGS_UDF_PTX_VERSION = (11, 8)
+
+def _get_cuda_version_from_ptx_file(path):
+    """
+    https://docs.nvidia.com/cuda/parallel-thread-execution/
+    Each PTX module must begin with a .version
+    directive specifying the PTX language version
+
+    example header:
+    //
+    // Generated by NVIDIA NVVM Compiler
+    //
+    // Compiler Build ID: CL-31057947
+    // Cuda compilation tools, release 11.6, V11.6.124
+    // Based on NVVM 7.0.1
+    //
+
+    .version 7.6
+    .target sm_52
+    .address_size 64
+
+    """
+    with open(path) as ptx_file:
+        for line in ptx_file:
+            if line.startswith(".version"):
+                ver_line = line
+                break
+        else:
+            raise ValueError("Could not read CUDA version from ptx file.")
+    version = ver_line.strip("\n").split(" ")[-1]
+    ptx_major = version.split(".")[0]
+    ptx_minor = version.split(".")[1]
+
+    # from ptx_docs/release_notes above:
+    maj_ver_map = {"7": "11", "8": "12"}
+    cuda_major = maj_ver_map.get(ptx_major)
+    if cuda_major is None:
+        raise ValueError(
+            f"Could not map major PTX version {ptx_major} to a CUDA version"
+        )
+
+    return int(cuda_major), int(ptx_minor)
```

Comment:
Are we guaranteed that the PTX minor version matches the CUDA minor version?

Reply:
In the docs this pattern holds for the cuda versions we support but doesn't seem to be guaranteed anywhere. While thinking about this I also realized we might be better off just mapping directly from the full version string in the file to a full cuda version to avoid confusing things if the pattern does not hold in the future. I made some changes in https://github.com/rapidsai/cudf/pull/12534/commits/49a210eba43b0c7e4414318d5a0490d4bfd07905, what do you think?

Reply:
Seems fine to me! Fewer assumptions is better. :)

Do we need mappings for CUDA 11.0, 11.1? We don't support those in RAPIDS anymore. We also require CUDA 11.5 to build cudf, so if the PTX version is guaranteed to be at least the minimum version to build, we could drop 11.2-11.4 as well.

Reply:
Dropped those ones!